File: solutions/scala/125-employee-free-time.scala
/**
 * Problem 125: Employee Free Time (LeetCode 759)
 * Difficulty: Hard
 * Language: Scala
 */
def employeeFreeTime(schedules: List[List[(Int, Int)]]): List[(Int, Int)] = {
  val sorted = schedules.flatten.sortBy(_._1)
  val merged = sorted.foldLeft(List.empty[(Int, Int)]) {
    case ((a, b) :: t, (s, e)) if b >= s => (a, b max e) :: t
    case (acc, iv) => iv :: acc
  }.reverse
  merged.sliding(2).collect {
    case List((_, e1), (s2, _)) if e1 < s2 => (e1, s2)
  }.toList
}

File: solutions/rust/125-employee-free-time.rs
/**
 * Problem 125: Employee Free Time (LeetCode 759)
 * Difficulty: Hard
 * Language: Rust
 */
fn employee_free_time(schedules: Vec<Vec<(i32, i32)>>) -> Vec<(i32, i32)> {
    let mut all: Vec<(i32, i32)> = schedules.into_iter().flatten().collect();
    all.sort();
    let merged = all.into_iter().fold(vec![], |mut acc: Vec<(i32,i32)>, (s, e)| {
        if let Some(last) = acc.last_mut() {
            if last.1 >= s { last.1 = last.1.max(e); return acc; }
        }
        acc.push((s, e)); acc
    });
    merged.windows(2).filter_map(|w| {
        if w[0].1 < w[1].0 { Some((w[0].1, w[1].0)) } else { None }
    }).collect()
}
